### PR TITLE
Add zero padding for matmul ops - Take 2

### PR DESCRIPTION
**Pull request by nsorabaTT**
This PR is a requeue with a small diff from the previous PR, which was reverted. Please refer to that for complete context.
https://github.com/tenstorrent/tt-metal/pull/22333

### The incremental change from the previous PR
MNIST has three matmuls:
| Reduction dim | Padding | Padding % |
|--------------|---------|-----------|
| 784          | 16      | 2%        |
| 120          | 8       | 6.6%      |
| 84           | 12      | 14.2%     |

There are 3 layers in this MNIST test:
```python
self.fc1 = torch.nn.Linear(784, 120)
self.fc2 = torch.nn.Linear(120, 84)
self.fc3 = torch.nn.Linear(84, 10)
```
Collectively, the the high percentage of padding adds 8-12% extra time. Thus, for these really small inputs that complete in a few microseconds, the padding overhead is high and causes a dip in performance.

Note that without padding, it is not guaranteed that these ops will produce correct results.

We update performance metrics for MNIST tests, reducing the expected performance by 9%.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passing https://github.com/tenstorrent/tt-metal/actions/runs/15423508215
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passing: https://github.com/tenstorrent/tt-metal/actions/runs/15423510891/job/43404980766
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI Passing https://github.com/tenstorrent/tt-metal/actions/runs/15434256620
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passing: https://github.com/tenstorrent/tt-metal/actions/runs/15423518224/job/43405020992
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [x] New/Existing tests provide coverage for changes
- [x] T300K model perf tests https://github.com/tenstorrent/tt-metal/actions/runs/15425331229

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp [hunk 1]
```diff
@@ -91,6 +95,15 @@ void kernel_main() {
             // Now we have the block in the CB address, we can mcast to dests!
             uint64_t in0_multicast_data_addr = in0_multicast_data_noc | in0_start_address;
 
+            // Zero out padded regions for the very last tile
+            if constexpr (in0_last_ktile_w > 0) {
+                DPRINT << "is_last_ktile_padded: " << (uint32_t)is_last_ktile_padded << ENDL();
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in0_sender_dram_sharded.cpp]: Please remove DPRINT.

**Review by bbradelTT — APPROVED** [review]

**Review by esmalTT — APPROVED** [review]
